### DISCUSSION TITLE
Converting a list of objects to YAML document ('---') with cue export

### DISCUSSION THREAD

**Discussion by PierreR** [cuelang/cue#996] [Q&A]
Is there a way to convert a list of objects to a '---'-separated stream of YAML values with `cue export --out yaml` ?

Or do I always need to use `MarshalStream` together with `cue cmd`

**Comment by seh** [comment]
I was getting ready to ask the same question, as, to my surprise, this desire has come up several times over the last week.

I don't think using [`yaml.MarshalStream`](https://pkg.go.dev/cuelang.org/go@v0.3.2/pkg/encoding/yaml#MarshalStream) necessitates using _cue cmd_, though; you can emit a top-level string value and request the output format be "text." What I can't remember, though, is whether this works without use of the `--expression` flag. I found that trying to emit a top-level string value didn't work, but emitting a struct with a field whose value is of type string works in concert with `--expression` to select that field's value.

**Reply by myitcv** [reply]
@seh I think you've actually hit upon a bug here. AFAIU `-e` should understand how to resolve builtins:

```
exec cue export --out yaml -e 'yaml.MarshalStream(X)' x.cue

-- x.cue --
X: [
	{
		a: 1
	},
	{
		b: 2
	},
]
```

gives:

```
reference "yaml" not found:
    --expression:1:1
```

Raised as https://github.com/cuelang/cue/issues/999

**Reply by PierreR** [reply]
@myitcv  I am wondering if there is not another bug. Shouldn't `yaml.MarshalStream` insert `---` at the beginning of the yaml file ? It only inserts `---` between each elements of the list.

**Reply by myitcv** [reply]
I'm unclear on that, @PierreR. Please feel free to open an issue, ideally with a reference that explains why `---` should appear at the start. My understanding was that `---` was used to _separate_ objects, but I could well be wrong.

**Reply by seh** [reply]
There's no document separator (["directives end marker"](https://yaml.org/spec/1.2/spec.html#id2800401)) required at the start of the first YAML document in a stream if there is no [document prefix](https://yaml.org/spec/1.2/spec.html#id2800168) preceding the document itself.

**Reply by seh** [reply]
Paul, the bug I was describing is different from what you attempted there with the `--expression` flag. Instead, it was this situation:
```cue
import "encoding/yaml"

let objs = [
	for v in [1, 2, 3] {
		input:  v
		double: v * 2
	},
]

yaml.MarshalStream(objs)
```
You can then invoke _cue export stream.cue --out text_ and get a YAML document stream. While that small example works fine for me today, I've run into larger examples where I couldn't get it to work—unless I nested that call to `yaml.MarshalStream` inside of a field that I then plucked out with `--expression`.

**Reply by myitcv** [reply]
@seh - thanks for clarifying. If you do have a repro of that, that's also worthy of an issue :)

**Reply by myitcv** [reply]
> There's no document separator (["directives end marker"](https://yaml.org/spec/1.2/spec.html#id2800401)) required at the start of the first YAML document in a stream if there is no [document prefix](https://yaml.org/spec/1.2/spec.html#id2800168) preceding the document itself.

Thanks for providing references.

Just to be clear: I'm not ruling out there being bugs here, rather I was keen to see examples of bugs in the light of spec references like this (with the wrinkle that some bugs are dependent on the version of a spec... something we might well need to handle in the future in any case)

**Reply by seh** [reply]
This problem with trying to use `--out text` _without_ the `--expression` flag bit me again today.

If my top-level value is `yaml.MarshalStream(v)` and I run _cue eval file --out text_, it fails with "incomplete value." If instead I revise that to be `x: yaml.MarshalStream(v)` and run _cue eval file --out text --expression x_, it works as expected.

**Reply by myitcv** [reply]
@seh please can you provide a repro?

**Reply by myitcv** [reply]
(grrr that sent too early). 

Because I seemingly can't reproduce the issue you are referring to.

**Reply by seh** [reply]
I'm not sure if the previous time this happened it was due to the same cause, but later yesterday I realized that I made a mistake: I was trying to export two different top-level string values _from the same package_. The error message didn't make clear that that was the problem. By nesting that string value inside of a field, that then revealed an error about trying to unify a struct and a string, which tipped me off further.

**Reply by myitcv** [reply]
Got it, that makes sense. If you are able to provide a repro for the poor error message, they're always good to fix.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/996.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
